### PR TITLE
Track airtime

### PR DESCRIPTION
**Pull request by Nicolasdejean**
#65 #30

### CODE REVIEW

File: examples/ABP/ABP.ino [hunk 1]
```diff
@@ -8,7 +8,7 @@ const byte appSKey[16] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/ABP/ABP.ino]: Use original line

File: examples/DeviceInfo/DeviceInfo.ino [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/DeviceInfo/DeviceInfo.ino]: Use original line

File: examples/QuickStart/QuickStart.ino [hunk 3]
```diff
@@ -7,7 +7,7 @@ const byte appKey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/QuickStart/QuickStart.ino]: Use original line

File: examples/Receive/Receive.ino [hunk 4]
```diff
@@ -7,7 +7,7 @@ const byte appKey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/Receive/Receive.ino]: Use original line

File: examples/Send/Send.ino [hunk 5]
```diff
@@ -7,7 +7,7 @@ const byte appKey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/Send/Send.ino]: Use original line

File: examples/Sensors/DHT/DHT.ino [hunk 6]
```diff
@@ -19,7 +19,7 @@ const byte appKey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 
 DHT dht(DHTPIN, DHTTYPE);
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/Sensors/DHT/DHT.ino]: Use original line

File: examples/Sensors/LightSensor/LightSensor.ino [hunk 7]
```diff
@@ -10,7 +10,7 @@ const byte appKey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/Sensors/LightSensor/LightSensor.ino]: Use original line

File: examples/Workshop/Workshop.ino [hunk 8]
```diff
@@ -20,7 +20,7 @@ const byte appSKey[16] = { ... };
 #define loraSerial Serial1
 #define debugSerial Serial
 
-TheThingsNetwork ttn(loraSerial, debugSerial, /* TTN_FP_EU868 or TTN_FP_US915 */);
+TheThingsNetwork ttn(Stream& modemStream, Stream& debugStream, /* TTN_FP_EU868 or TTN_FP_US915 */);
```
Comment (johanstokking) [inline examples/Workshop/Workshop.ino]: Use original line

File: keywords.txt [hunk 9]
```diff
@@ -27,11 +27,11 @@ poll                    KEYWORD2
...
-TTN_FP_US915            LITERAL1
\ No newline at end of file
+TTN_DEFAULT_WAIT_TIME   KEYWORD2
+TTN_DEFAULT_SF          KEYWORD2
+TTN_DEFAULT_FSB         KEYWORD2
+TTN_ADR_SUPPORTED       KEYWORD2
+TTN_PWRIDX_868          KEYWORD2
+TTN_PWRIDX_915          KEYWORD2
+TTN_FP_EU868            KEYWORD2
+TTN_FP_US915            KEYWORD2
```
Comment (johanstokking) [inline keywords.txt]: Aren't these literals? @FokkeZB

File: src/TheThingsNetwork.cpp [hunk 10]
```diff
@@ -234,6 +244,47 @@ int TheThingsNetwork::poll(int port, bool confirm) {
...
+      return (i - 4);
+    i = i + 1;
+  }
+  i = -1;
+  while (message[++i]) {
+    message[i] >= '0' && message[i] <= '9' ? stock = (stock + (message[i] - 48)) * 10 : stock = stock;
+  }
+  stock == 0 && message == F("on") ? stock = 1 : stock = stock / 10;
+  return (stock);
+}
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: What is `getInfo()` doing?

File: src/TheThingsNetwork.cpp [hunk 11]
```diff
@@ -234,6 +244,47 @@ int TheThingsNetwork::poll(int port, bool confirm) {
...
+    if (str == message)
+      return (i - 4);
+    i = i + 1;
+  }
+  i = -1;
+  while (message[++i]) {
+    message[i] >= '0' && message[i] <= '9' ? stock = (stock + (message[i] - 48)) * 10 : stock = stock;
+  }
+  stock == 0 && message == F("on") ? stock = 1 : stock = stock / 10;
+  return (stock);
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: Remove parentheses

File: src/TheThingsNetwork.cpp [hunk 12]
```diff
@@ -234,6 +244,47 @@ int TheThingsNetwork::poll(int port, bool confirm) {
...
+    str.concat(i);
+    if (str == message)
+      return (i - 4);
+    i = i + 1;
+  }
+  i = -1;
+  while (message[++i]) {
+    message[i] >= '0' && message[i] <= '9' ? stock = (stock + (message[i] - 48)) * 10 : stock = stock;
+  }
+  stock == 0 && message == F("on") ? stock = 1 : stock = stock / 10;
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: Use if-statement; use `?:` for returning values

File: src/TheThingsNetwork.cpp [hunk 13]
```diff
@@ -234,6 +244,47 @@ int TheThingsNetwork::poll(int port, bool confirm) {
...
+    str = "";
+    str.concat(F("4/"));
+    str.concat(i);
+    if (str == message)
+      return (i - 4);
+    i = i + 1;
+  }
+  i = -1;
+  while (message[++i]) {
+    message[i] >= '0' && message[i] <= '9' ? stock = (stock + (message[i] - 48)) * 10 : stock = stock;
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: Use if-statement; `stock = stock` doesn't make sense

File: src/TheThingsNetwork.cpp (hunk 10 above)
Comment (Nicolasdejean) [inline src/TheThingsNetwork.cpp]: @johanstokking It takes only the numbers from the ``` String message ``` so I can do calculation in ``` trackAirtime() ``` for example if message receives ``` sf10 ``` it takes only ``` 10 ``` and stock it in ``` this->info.sf ```.

Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: OK, parse it using the pattern `SFxxBWyyy` where `xx` is either 1 or 2 digits.

What is that `"on"` doing?

**Pull request closed without merging**